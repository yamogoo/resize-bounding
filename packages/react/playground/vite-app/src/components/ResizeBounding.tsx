import ResizeBoundingComponent, { type Props } from "react-resize-bounding";

const ResizeBounding = (props: Props) => {
  return (
    <ResizeBoundingComponent
      {...props}
      options={{
        knob: {
          show: true,
        },
      }}
    >
      {props.children}
    </ResizeBoundingComponent>
  );
};

export default ResizeBounding;
