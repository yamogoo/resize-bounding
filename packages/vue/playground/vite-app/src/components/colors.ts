export const colors = {
  backgroundApp: "#171717",
  backgroundPrimary: "#2B2B2B",
  backgroundSecondary: "#171717",
  foregroundPrimary: "#5a5a5a",
  foregroundSecondary: "#C0C0C0",
  borderNormal: "#313131",
  accent: "#776AFF",
  resizerKnobNormal: "#E1E1E1",
};
