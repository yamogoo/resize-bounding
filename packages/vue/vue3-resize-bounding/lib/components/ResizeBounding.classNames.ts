import { mergeStyleSets } from "@fluentui/merge-styles";
import type {
  IStyles,
  IResizeBoundingClassNames,
  Options,
} from "../shared/typings";

export const PREFIX = "resize-bounding-";

export const defaultOptions: Options = {
  prefix: PREFIX,
  width: 4,
  activeAreaWidth: undefined,
  position: "central",
  knob: {
    show: false,
    normalHidden: false,
  },
  cursor: {
    horizontal: "col-resize",
    vertical: "row-resize",
  },
  touchActions: true,
};

export const globalClassNames = (prefix = PREFIX) => ({
  container: `${prefix}container`,
  pane: `${prefix}pane`,
  splitter: `${prefix}splitter`,
  splitterContainer: `${prefix}splitter--container`,
  knob: `${prefix}knob`,
});

export const defaultStyles = (prefix: string): IStyles => ({
  container: [
    globalClassNames(prefix).container,
    { displayName: globalClassNames(prefix).container, position: "relative" },
  ],
  pane: [
    globalClassNames(prefix).pane,
    {
      displayName: globalClassNames(prefix).pane,
      position: "absolute",
      display: "block",
      zIndex: 9999,
      touchAction: "none",
    },
  ],
  splitter: [
    globalClassNames(prefix).splitter,
    {
      displayName: globalClassNames(prefix).splitter,
      position: "absolute",
      zIndex: 9999,
      transition: "background 125ms ease-out",
      [`.${globalClassNames(prefix).pane}.focused &, .${globalClassNames(prefix).pane}.pressed &`]:
        {
          background: "cornflowerblue",
        },
    },
  ],
  splitterContainer: [
    globalClassNames(prefix).splitterContainer,
    {
      displayName: globalClassNames(prefix).splitterContainer,
      position: "relative",
      top: "50%",
      left: "50%",
      width: `0px`,
      height: `0px`,
    },
  ],
  knob: [
    globalClassNames(prefix).knob,
    {
      displayName: globalClassNames(prefix).knob,
      position: "relative",
      width: "64px",
      height: "8px",
      background: "gray",
      borderRadius: "4px",
      transform: "translate(-50%, -50%)",
      transition: "background 125ms ease-out",
      [`.${globalClassNames(prefix).pane}.focused &,
        .${globalClassNames(prefix).pane}.pressed &`]: {
        background: "cornflowerblue",
      },
    },
  ],
});

export const getClassNames = (
  args: Partial<IStyles>,
  prefix = PREFIX,
): IResizeBoundingClassNames => {
  return mergeStyleSets(defaultStyles(prefix), args);
};
