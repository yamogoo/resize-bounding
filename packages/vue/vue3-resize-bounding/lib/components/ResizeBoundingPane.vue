<template>
  <div
    ref="refPane"
    data-testid="resize-bounding-pane"
    :class="[
      classNames.pane,
      { active: isFocused || isPressed },
      options.addStateClasses && {
        normal: !isPressed && !isFocused,
        focused: isFocused,
        pressed: isPressed,
      },
    ]"
    :style="[paneComputedStyle]"
  >
    <div
      data-testid="resize-bounding-splitter"
      :class="[classNames.splitter]"
      :style="[splitterComputedStyle]"
    >
      <div
        v-if="
          (isFocused || !options?.knob?.normalHidden) && options?.knob?.show
        "
        data-testid="resize-bounding-splitter-container"
        :class="[classNames.splitterContainer]"
        :style="[containerComputedStyles]"
      >
        <div data-testid="resize-bounding-knob" :class="[classNames.knob]">
          <slot v-if="$slots.default"></slot>
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import {
  onMounted,
  onUnmounted,
  ref,
  type Ref,
  computed,
  type HTMLAttributes,
} from "vue";

import {
  SplitterPositions,
  type IStyles,
  type Options,
  type IResizeBoundingClassNames,
  PaneDirectionAliases,
} from "../shared/typings";

import {
  paneBaseStyles,
  splitterBaseStyles,
} from "./ResizeBoundingPane.styles";

const props = withDefaults(defineProps<Props>(), {
  direction: PaneDirections.RIGHT,
});

const emits = defineEmits<{
  (e: Emits.DRAG_START, data: PaneEmittedData): void;
  (e: Emits.DRAG_MOVE, data: PaneEmittedData): void;
  (e: Emits.DRAG_END, data: PaneEmittedData): void;
  (e: Emits.FOCUS, isFocused: boolean): void;
}>();

const refPane: Ref<HTMLDivElement | null> = ref(null);

const isFocused = ref(false),
  isPressed = ref(false);

let isResizing = false;

const paneComputedStyle = computed(() => {
  const _width = props.options?.width ?? 1;

  if (refPane.value && _width) {
    const _areaWidth = props.options.activeAreaWidth ?? props.options.width;
    const _styles = paneBaseStyles(
      _width,
      _areaWidth,
      props.options?.position ?? SplitterPositions.CENTER,
    );
    const value: HTMLAttributes["style"] =
      _styles[props.direction as PaneDirections];
    return value;
  }
  return;
});

const splitterComputedStyle = computed(() => {
  const _width = props.options?.width;

  if (refPane.value && _width) {
    const _areaWidth = props.options.activeAreaWidth ?? props.options.width;

    const _styles = splitterBaseStyles(_width, _areaWidth);
    const value: HTMLAttributes["style"] =
      _styles[props.direction as PaneDirections];
    return value;
  }
});

const containerComputedStyles = computed(() => {
  const isHorizontal = checkIsHorizontal(props.direction);
  return { transform: `rotate(${isHorizontal ? 90 : 0}deg)` };
});

const updateCursor = (state: boolean) => {
  let cursorActive: string;

  if (checkIsHorizontal(props.direction)) {
    cursorActive = props.options?.cursor?.horizontal ?? "auto";
  } else if (checkIsVertical(props.direction)) {
    cursorActive = props.options?.cursor?.vertical ?? "auto";
  } else cursorActive = "auto";

  if (refPane.value) refPane.value.style.cursor = state ? cursorActive : "auto";
};

const emitFocus = (state: boolean) => {
  emits(Emits.FOCUS, state);
};

const onFocus = (e: PointerEvent, isFocused: boolean): void => {
  e.stopPropagation();
  onSelected(isFocused);
};

const onSelected = (state: boolean): void => {
  isFocused.value = state;
  updateCursor(state);
  emitFocus(state);
};

const onDragStart = (e: PointerEvent): void => {
  e.preventDefault();
  e.stopImmediatePropagation();

  if (!props.options.touchActions && e.pointerType === "touch") return;

  isResizing = true;
  isPressed.value = true;

  onSelected(isResizing);

  const el = e.currentTarget as HTMLDivElement;
  el.setPointerCapture(e.pointerId);

  emits(Emits.DRAG_START, {
    x: Math.round(e.clientX),
    y: Math.round(e.clientY),
    dir: props.direction,
  });

  const onDragMove = (e: PointerEvent): void => {
    emits(Emits.DRAG_MOVE, {
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: props.direction,
    });
  };

  const onDragEnd = (e: PointerEvent): void => {
    isResizing = false;
    isPressed.value = false;

    updateCursor(isResizing);

    const el = e.currentTarget as HTMLDivElement;
    el.releasePointerCapture(e.pointerId);

    el.removeEventListener("pointermove", onDragMove);
    el.removeEventListener("pointerup", onDragEnd);

    emits(Emits.DRAG_END, {
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: props.direction,
    });
  };

  el.addEventListener("pointermove", onDragMove);
  el.addEventListener("pointerup", onDragEnd);
};

const onDragCancel = (e: PointerEvent): void => {
  const el = e.currentTarget as HTMLDivElement;

  isResizing = false;
  el.releasePointerCapture(e.pointerId);

  emits(Emits.DRAG_END, {
    x: Math.round(e.clientX),
    y: Math.round(e.clientY),
    dir: props.direction,
  });
};

const addEventListeners = () => {
  const el = refPane.value;

  if (el) {
    el.addEventListener("pointerenter", (e) => onFocus(e, true));
    el.addEventListener("pointerleave", (e) => onFocus(e, false));

    el.addEventListener("pointerdown", onDragStart);
    el.addEventListener("pointercancel", onDragCancel);
  }
};

const removeEventListeners = () => {
  const el = refPane.value;

  if (el) {
    el.removeEventListener("pointerenter", (e) => onFocus(e, true));
    el.removeEventListener("pointerleave", (e) => onFocus(e, false));

    el.removeEventListener("pointerdown", onDragStart);
    el.addEventListener("pointercancel", onDragCancel);
  }
};

onMounted(addEventListeners);
onUnmounted(removeEventListeners);

defineExpose({ refPane });
</script>

<script lang="ts">
import { PaneDirections, type PaneDirectionKey } from "../shared/typings";

export interface Props {
  prefix: string;
  direction: PaneDirectionKey;
  options: Options;
  classNames: IResizeBoundingClassNames;
  styles?: Partial<IStyles>;
}

export enum Emits {
  FOCUS = "focus",
  DRAG_START = "drag:start",
  DRAG_MOVE = "drag:move",
  DRAG_END = "drag:end",
}

export interface PaneEmittedData {
  x: number;
  y: number;
  dir: string;
}

const checkIsHorizontal = (direction: string): boolean =>
  new RegExp(
    `[${PaneDirections.LEFT} | ${PaneDirections.RIGHT} | ${PaneDirectionAliases.HORIZONTAL}]`,
  ).test(direction);
const checkIsVertical = (direction: string): boolean =>
  new RegExp(
    `[${PaneDirections.TOP} | ${PaneDirections.BOTTOM} | ${PaneDirectionAliases.VERTICAL}]`,
  ).test(direction);
</script>
