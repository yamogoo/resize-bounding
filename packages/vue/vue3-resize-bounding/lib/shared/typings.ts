import fluentui from "@fluentui/merge-styles";

export interface Knob {
  show: boolean;
  normalHidden: boolean;
}

export enum SplitterPositions {
  CENTER = "central",
  EXTERNAL = "external",
  INTERNAL = "internal",
}

export type PanePosition = SplitterPositions | string;

export interface Options {
  prefix: string;
  width: number;
  activeAreaWidth?: number;
  position: PanePosition;
  cursor: Partial<
    Record<"horizontal" | "vertical", CSSStyleDeclaration["cursor"]>
  >;
  knob: Partial<Knob>;
  touchActions: boolean;
}

export type ResizeBoundingClassNames =
  | "container"
  | "pane"
  | "splitter"
  | "splitterContainer"
  | "knob";

export interface IResizeBoundingClassNames {
  container: string;
  pane: string;
  splitter: string;
  splitterContainer: string;
  knob: string;
}

export interface IStyles extends Record<ResizeBoundingClassNames, IStyle> {}
export type IStyle = fluentui.IStyle;
export type IStyleSet = fluentui.IStyleSet;

/* * * Pane * * */

export enum PaneDirectionAliases {
  HORIZONTAL = "h",
  VERTICAL = "v",
}

export enum PaneDirections {
  LEFT = "l",
  RIGHT = "r",
  BOTTOM = "b",
  TOP = "t",
}
