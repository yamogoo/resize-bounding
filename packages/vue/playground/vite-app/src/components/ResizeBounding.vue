<template>
  <ResizeBounding
    :options="{
      knob: {
        show: true,
      },
    }"
  >
    <slot></slot>
  </ResizeBounding>
</template>

<script setup lang="ts">
import ResizeBounding from "vue3-resize-bounding";
</script>
