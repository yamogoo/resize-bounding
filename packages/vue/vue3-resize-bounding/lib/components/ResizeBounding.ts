import { IStyles, Options, type PaneDirectionKey } from "../shared/typings";

export interface Props
  extends Partial<{
    width: number;
    height: number;
    minWidth: number;
    maxWidth: number;
    minHeight: number;
    maxHeight: number;
    disabled: boolean;
    directions: PaneDirectionKey;
    alwaysShowKnob: boolean;
    options: Partial<Options>;
    styles: Partial<IStyles>;
    updateWidth: (width: number) => void;
    updateHeight: (height: number) => void;
    dragStart: (dir: string) => void;
    dragMove: (dir: string) => void;
    dragEnd: (dir: string) => void;
    focus: (data: { state: boolean; direction: string }) => void;
  }> {}

export enum Emits {
  FOCUS = "focus",
  UPDATE_WIDTH = "update:width",
  UPDATE_HEIGHT = "update:height",
  DRAG_START = "drag:start",
  DRAG_MOVE = "drag:move",
  DRAG_END = "drag:end",
}
