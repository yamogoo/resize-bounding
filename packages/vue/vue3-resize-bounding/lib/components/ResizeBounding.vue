<template>
  <div
    ref="refRoot"
    data-testid="resize-bounding-container"
    :class="[classNames.container, { disabled }]"
    :style="[computedStyle]"
  >
    <slot></slot>
    <template v-for="{ show, direction } in panes" :key="direction">
      <ResizeBoundingPane
        v-if="!disabled && show"
        :prefix="options.prefix ?? ''"
        :direction="direction"
        :options="options"
        :styles
        :classNames
        @focus="
          (isFocused) => {
            $emit(Emits.FOCUS, { state: isFocused, direction });
          }
        "
        @drag:start="onDragStart"
        @drag:move="onDragMove"
        @drag:end="onDragEnd"
      >
        <template #default>
          <slot name="knob"></slot>
        </template>
      </ResizeBoundingPane>
    </template>
  </div>
</template>

<script setup lang="ts">
import { ref, computed, type ComputedRef, type HTMLAttributes } from "vue";
import deepmerge from "deepmerge";

import { type Props, Emits } from "./ResizeBounding";
import { PaneDirections, PaneDirectionAliases } from "../shared/typings";

import ResizeBoundingPane, {
  type PaneEmittedData,
} from "./ResizeBoundingPane.vue";

import {
  defaultOptions,
  getClassNames,
  PREFIX,
} from "./ResizeBounding.classNames";

const props = withDefaults(defineProps<Props>(), {
  minWidth: 0,
  minHeight: 0,
});

const emits = defineEmits<{
  (e: Emits.UPDATE_WIDTH, width: number): void;
  (e: Emits.UPDATE_HEIGHT, height: number): void;
  (e: Emits.DRAG_START, dir: string): void;
  (e: Emits.DRAG_MOVE, dir: string): void;
  (e: Emits.DRAG_END, dir: string): void;
  (e: Emits.FOCUS, data: { state: boolean; direction: string }): void;
}>();

const refRoot = ref<HTMLDivElement | null>(null);

let { width: newWidth, height: newHeight } = props;
let prevWidth = newWidth,
  prevHeight = newHeight;

const options = computed(() => {
  return deepmerge(defaultOptions, props.options ?? {});
});

const classNames = getClassNames(
  props.styles ?? {},
  options.value.prefix ?? PREFIX,
);

const computedStyle: ComputedRef<HTMLAttributes["style"]> = computed(() => {
  const _width = {
    width: `${props.width}px`,
    minWidth: `${props.width}px`,
  };

  const _height = {
    height: `${props.height}px`,
    minHeight: `${props.height}px`,
  };

  return {
    ...(props.width && _width),
    ...(props.height && _height),
  };
});

const getDirectionAlias = (d: PaneDirections): PaneDirectionAliases => {
  if (d === PaneDirections.LEFT || d === PaneDirections.RIGHT)
    return PaneDirectionAliases.HORIZONTAL;
  return PaneDirectionAliases.VERTICAL;
};

const panes = computed(() => {
  const directions = props.directions ?? "";

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  return Object.entries(PaneDirections).map(([_, v]) => {
    return {
      show:
        RegExp(v).test(directions) ||
        RegExp(getDirectionAlias(v)).test(directions),
      direction: v,
    };
  });
});

let startWidth = props.width ?? 0,
  startHeight = props.height ?? 0;

let startX = 0,
  startY = 0;

const onDragStart = ({ x, y, dir }: PaneEmittedData): void => {
  startWidth = props.width ?? 0;
  startHeight = props.height ?? 0;

  startX = x;
  startY = y;

  emits(Emits.DRAG_START, dir);
};

const truncateInRange = (
  min: number,
  max: number | undefined,
  next: number,
): number => {
  const _max = max ?? Number.POSITIVE_INFINITY;

  if (next <= min) return min;
  else if (next >= _max) return _max;
  return next;
};

const onDragMove = ({ x, y, dir }: PaneEmittedData): void => {
  if (!refRoot.value) return;

  emits(Emits.DRAG_MOVE, dir);

  if (dir === PaneDirections.LEFT) {
    newWidth = startWidth + (startX - x);
    if (prevWidth === newWidth) return;

    const truncated = truncateInRange(props.minWidth, props.maxWidth, newWidth);
    emits(Emits.UPDATE_WIDTH, truncated);
    prevWidth = truncated;
  } else if (dir === PaneDirections.RIGHT) {
    newWidth = startWidth + (x - startX);

    if (prevWidth === newWidth) return;

    const truncated = truncateInRange(props.minWidth, props.maxWidth, newWidth);
    emits(Emits.UPDATE_WIDTH, truncated);
    prevWidth = truncated;
  } else if (dir === PaneDirections.TOP) {
    newHeight = startHeight + (startY - y);

    if (prevHeight === newHeight) return;

    const truncated = truncateInRange(
      props.minHeight,
      props.maxHeight,
      newHeight,
    );

    emits(Emits.UPDATE_HEIGHT, truncated);
    prevHeight = truncated;
  } else if (dir === PaneDirections.BOTTOM) {
    newHeight = startHeight + (y - startY);

    if (prevHeight === newHeight) return;

    const truncated = truncateInRange(
      props.minHeight,
      props.maxHeight,
      newHeight,
    );

    emits(Emits.UPDATE_HEIGHT, truncated);
    prevHeight = truncated;
  }
};

const onDragEnd = ({ dir }: PaneEmittedData): void => {
  emits(Emits.DRAG_END, dir);
};
</script>
