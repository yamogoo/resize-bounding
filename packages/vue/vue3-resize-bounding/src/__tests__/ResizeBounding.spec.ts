import { describe, test, expect, beforeEach, vi } from "vitest";
import { mount, shallowMount, VueWrapper } from "@vue/test-utils";
import type { HTMLAttributes } from "vue";

import { DataTestIds, PREFIX } from "./setup";

import ResizeBounding from "../../lib/components/ResizeBounding.vue";
import Vue3BbResizePane, {
  type PaneEmittedData,
} from "../../lib/components/ResizeBoundingPane.vue";

import { Emits } from "../../lib/components/ResizeBounding";
import { PaneDirectionAliases, PaneDirections } from "../../lib/shared/typings";

describe("ResizeBounding", () => {
  beforeEach(() => {
    Element.prototype.setPointerCapture = vi.fn();
    Element.prototype.releasePointerCapture = vi.fn();
  });

  describe("components", () => {
    test.each([
      ...Object.values(PaneDirections),
      ...Object.values(PaneDirectionAliases),
    ])(
      `Should not display any selected borders when set to "disabled" (--%s)`,
      (direction) => {
        const wrapper = mount(ResizeBounding, {
          props: {
            directions: direction,
            disabled: true,
          },
        });

        const panEl = wrapper.findAll(`[data-testid="${DataTestIds.PANE}"]`);

        expect(panEl.length).toBe(0);
        expect(panEl.length).toMatchSnapshot();
      },
    );

    test.each(["rl", "ltrb", "bltr", "bl", "tr"])(
      "should render %s panes (panes computed value)",
      (directions) => {
        const wrapper = mount(ResizeBounding, {
          props: {
            directions,
          },
        });

        const panEls = wrapper.findAll(`[data-testid="${DataTestIds.PANE}"]`);
        expect(panEls.length).toBe(directions.length);
        expect(panEls.length).toMatchSnapshot();
      },
    );

    test.each(["sd", "wo", ""])(
      "should not render any pane element",
      (directions) => {
        const wrapper = mount(ResizeBounding, {
          props: {
            directions,
          },
        });

        const panEls = wrapper.findAll(`[data-testid="${DataTestIds.PANE}"]`);
        expect(panEls.length).toBe(0);
        expect(panEls.length).toMatchSnapshot();
      },
    );
  });

  describe("slots", () => {
    test.each(["<p>Inner Container</p>"])(
      "should render default slot (%s)",
      (slot) => {
        const wrapper = mount(ResizeBounding, {
          props: {
            directions: "",
          },
          slots: {
            default: slot,
          },
        });

        const html = wrapper.find(`[data-testid="${DataTestIds.ROOT}"]`).html();
        expect(html).toContain(slot);
        expect(html).toMatchSnapshot();
      },
    );

    test.each(["<p>knob</p>"])("should render default slot", (slot) => {
      const wrapper = mount(ResizeBounding, {
        props: {
          directions: "r",
          options: {
            knob: {
              show: true,
            },
          },
        },
        slots: {
          knob: slot,
        },
      });

      const text = wrapper.find(`[data-testid="${DataTestIds.KNOB}"]`).text();
      expect(text).toBe("knob");
      expect(text).toMatchSnapshot();
    });
  });

  describe("classes", () => {
    describe("prefix", () => {
      test.each(["my-prefix"])("should have custom prefix (%s)", (prefix) => {
        const wrapper = shallowMount(ResizeBounding, {
          props: {
            options: {
              prefix,
            },
          },
        });

        const rootEl = wrapper.find(`[data-testid="${DataTestIds.ROOT}"]`);
        const classes = rootEl.classes();

        expect(classes.toString()).toContain(prefix);
        expect(classes).toMatchSnapshot();
      });

      test("should have default prefix (%s)", () => {
        const wrapper = shallowMount(ResizeBounding);

        const rootEl = wrapper.find(`[data-testid="${DataTestIds.ROOT}"]`);
        const classes = rootEl.classes();

        expect(classes.toString()).toContain(PREFIX);
        expect(classes).toMatchSnapshot();
      });
    });
  });

  describe("events", () => {
    describe("drag", () => {
      const checkData = (
        eventName: string,
        wrapper: VueWrapper,
        direction: string,
      ): PaneEmittedData => {
        const ev = wrapper.emitted(eventName) as PaneEmittedData[][];
        const data = ev[0][0];
        expect(data).toBe(direction);
        return data;
      };

      test.each(Object.values(PaneDirections))(
        `should emit the "${Emits.DRAG_START} -> ${Emits.DRAG_MOVE} -> ${Emits.DRAG_END}" event, which takes a direction argument "%s"`,
        async (direction) => {
          const wrapper = mount(ResizeBounding, {
            props: {
              directions: direction,
            },
          });
          await wrapper.setProps({ directions: direction });

          const paneWrapper = wrapper.findComponent(Vue3BbResizePane);
          expect(paneWrapper.exists()).toBeTruthy();

          const pane = paneWrapper.find(
            `[data-testid="${DataTestIds.SPLITTER}"]`,
          );

          const events: PaneEmittedData[] = [];

          const makeTrigger = async (
            event: string,
            eventData: { clientX: number; clientY: number },
            emittedEventName: string,
          ) => {
            await pane.trigger(event, eventData);
            paneWrapper.emitted(emittedEventName) as PaneEmittedData[][];
            events.push(checkData(emittedEventName, wrapper, direction));
          };

          await makeTrigger(
            "pointerdown",
            { clientX: 1, clientY: 1 },
            Emits.DRAG_START,
          );
          await makeTrigger(
            "pointermove",
            { clientX: 2, clientY: 2 },
            Emits.DRAG_MOVE,
          );
          await makeTrigger(
            "pointerup",
            { clientX: 3, clientY: 3 },
            Emits.DRAG_END,
          );

          expect(events.length).toBe(3);
          expect(events).toMatchSnapshot();
        },
      );

      test.each([
        [
          {
            alias: PaneDirectionAliases.HORIZONTAL,
            directions: [PaneDirections.RIGHT, PaneDirections.LEFT],
          },
        ],
        [
          {
            alias: PaneDirectionAliases.VERTICAL,
            directions: [PaneDirections.TOP, PaneDirections.BOTTOM],
          },
        ],
      ])(
        `check aliases: should emit "${Emits.DRAG_START}" with direction argument "%o"`,
        async ({ alias, directions }) => {
          const wrapper = mount(ResizeBounding, {
            props: {
              directions: alias,
            },
          });

          const paneEls = wrapper.findAll(
            `[data-testid="${DataTestIds.SPLITTER}"]`,
          );

          directions.map(async (_d, idx) => {
            await paneEls[idx].trigger("pointerdown", {
              clientX: 1,
              clientY: 1,
            });
          });

          const ev = wrapper.emitted(Emits.DRAG_START) as PaneEmittedData[][];

          let _directions: Array<PaneDirections> = [];
          for (const direction of directions) _directions.push(direction);

          expect(ev.length).toBe(directions.length);
          expect(_directions).toMatchObject(directions);
          expect(_directions).toMatchSnapshot();
        },
      );
    });
  });

  describe("styles", () => {
    describe("container", () => {
      test.each([{ background: "blue", display: "flex" }])(
        "should apply inline styles (%s)",
        (container: HTMLAttributes["style"]) => {
          const wrapper = shallowMount(ResizeBounding, {
            props: {
              style: container,
            },
          });

          const containerEl = wrapper.find(
            `[data-testid="${DataTestIds.ROOT}"]`,
          );
          const styles = containerEl.attributes("style");

          expect(styles).toContain("display: flex");
          expect(styles).toContain("background: blue");
          expect(styles).toMatchSnapshot();
        },
      );
    });
  });
});
