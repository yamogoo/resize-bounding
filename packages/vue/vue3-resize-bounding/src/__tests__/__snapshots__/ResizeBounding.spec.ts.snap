// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResizeBounding > classes > prefix > should have custom prefix (my-prefix) 1`] = `
[
  "my-prefixcontainer",
  "my-prefixcontainer-5",
]
`;

exports[`ResizeBounding > classes > prefix > should have default prefix (%s) 1`] = `
[
  "resize-bounding-container",
  "resize-bounding-container-0",
]
`;

exports[`ResizeBounding > components > Should not display any selected borders when set to "disabled" (--b) 1`] = `0`;

exports[`ResizeBounding > components > Should not display any selected borders when set to "disabled" (--h) 1`] = `0`;

exports[`ResizeBounding > components > Should not display any selected borders when set to "disabled" (--l) 1`] = `0`;

exports[`ResizeBounding > components > Should not display any selected borders when set to "disabled" (--r) 1`] = `0`;

exports[`ResizeBounding > components > Should not display any selected borders when set to "disabled" (--t) 1`] = `0`;

exports[`ResizeBounding > components > Should not display any selected borders when set to "disabled" (--v) 1`] = `0`;

exports[`ResizeBounding > components > should not render any pane element 1`] = `0`;

exports[`ResizeBounding > components > should not render any pane element 2`] = `0`;

exports[`ResizeBounding > components > should not render any pane element 3`] = `0`;

exports[`ResizeBounding > components > should render bl panes (panes computed value) 1`] = `2`;

exports[`ResizeBounding > components > should render bltr panes (panes computed value) 1`] = `4`;

exports[`ResizeBounding > components > should render ltrb panes (panes computed value) 1`] = `4`;

exports[`ResizeBounding > components > should render rl panes (panes computed value) 1`] = `2`;

exports[`ResizeBounding > components > should render tr panes (panes computed value) 1`] = `2`;

exports[`ResizeBounding > events > drag > check aliases: should emit "drag:start" with direction argument "{ alias: 'h', directions: [ 'r', 'l' ] }" 1`] = `
[
  "r",
  "l",
]
`;

exports[`ResizeBounding > events > drag > check aliases: should emit "drag:start" with direction argument "{ alias: 'v', directions: [ 't', 'b' ] }" 1`] = `
[
  "t",
  "b",
]
`;

exports[`ResizeBounding > events > drag > should emit the "drag:start -> drag:move -> drag:end" event, which takes a direction argument "b" 1`] = `
[
  "b",
  "b",
  "b",
]
`;

exports[`ResizeBounding > events > drag > should emit the "drag:start -> drag:move -> drag:end" event, which takes a direction argument "l" 1`] = `
[
  "l",
  "l",
  "l",
]
`;

exports[`ResizeBounding > events > drag > should emit the "drag:start -> drag:move -> drag:end" event, which takes a direction argument "r" 1`] = `
[
  "r",
  "r",
  "r",
]
`;

exports[`ResizeBounding > events > drag > should emit the "drag:start -> drag:move -> drag:end" event, which takes a direction argument "t" 1`] = `
[
  "t",
  "t",
  "t",
]
`;

exports[`ResizeBounding > slots > should render default slot 1`] = `
"<div data-testid="resize-bounding-container" class="resize-bounding-container resize-bounding-container-0">
  <p>Inner Container</p>
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</div>"
`;

exports[`ResizeBounding > slots > should render default slot 2`] = `"knob"`;

exports[`ResizeBounding > styles > container > should apply custom styles ({ background: 'red' }) 1`] = `"width: 320px; min-width: 320px; height: 400px; min-height: 400px;"`;

exports[`ResizeBounding > styles > container > should apply inline styles ({ background: 'blue', display: 'flex' }) 1`] = `"background: blue; display: flex;"`;
